docs/buffers.md

```diff
@@ -0,0 +1,143 @@
+# Buffers
+
+Buffer handles automatic memory migration between the host and the device or between devices when the kernel needs to access the memory. This helps you avoid copying memory from the host to the device manually.
+
+A synopsis of buffer use is as follows:
+
+```cpp
+#define GMEM_DATA_VEC 8
+
+cl_platform_id platform;
+cl_int err;
+// General setups
+err = clGetPlatformIDs(1, &platform, NULL);
+err = clGetDeviceIDs(platform, CL_DEVICE_TYPE_ALL, 1, &device, NULL);
+cl_context context = clCreateContext(NULL, 1, &device, NULL, NULL, &err);
+cl_command_queue queue = clCreateCommandQueue(context, device, CL_QUEUE_PROFILING_ENABLE, &err);
+// aocx_size is the size of aocx file
+// aocx_contents is the content of aocx file
+cl_program program = clCreateProgramWithBinary(context, 1, &device, &aocx_size, (const unsigned char **)&aocx_contents, &binary_status, &err);
+cl_kernel kernel = clCreateKernel(program, "kernel_name", &err);
+
+// Host side data that contains the information needed for kernel execution
+unsigned int *host_data = (unsigned int *)acl_aligned_malloc(global_work_size*sizeof(unsigned int)*GMEM_DATA_VEC);
+
+// Set up the expectation for a buffer, but not actually allocating memory yet
+cl_mem mem = clCreateBuffer(context, CL_MEM_READ_WRITE, global_work_size*sizeof(unsigned int)*GMEM_DATA_VEC, NULL, &err);
+
+// Enqueue a memory transfer operation
+cl_event event = {};
+err = clEnqueueWriteBuffer(queue, mem, CL_FALSE, 0, global_work_size*sizeof(unsigned int)*GMEM_DATA_VEC, host_data, 0, NULL, &event);
+
+// Informs the kernel that one of its arguments is the buffer
+err = clSetKernelArg(kernel, 0, sizeof(cl_mem), &mem);
+
+// Submit kernel execution to queue
+err = clEnqueueNDRangeKernel(queue, kernel, 2, NULL, &global_work_size_2d[0], &local_work_size_2d[0], 0, NULL, &event);
+
+... // Block on kernel finish and clean up resources
+```
+
+## Summary
+
+* Buffers are not allocated on the device until the kernel executes, the `clEnqueueWriteBuffer` function is called, or `CL_MEM_COPY_HOST_PTR` flag is defined when calling the function `clCreateBuffer`.
+* Buffer handles memory migration between the host and device automatically.
+* SYCL* runtime typically calls OpenCL functions in a different order in comparison to OpenCL's preferred order. Specifically, SYCL* runtime calls functions in the order of `clCreateBuffer` > `clEnqueueWriteBuffer` > `clSetKernelArg` > `clEnqueueNDRangeKernel`. In OpenCL, the optimal calls are `clCreateBuffer` > `clSetKernelArg` > `clEnqueueWriteBuffer` >  `clEnqueueNDRangeKernel`. The difference in order of calls sometimes leads to an unexpected result (eg. extra host to device memory copy), but you can avoid it if the `clSetKernelArg` function does not modify `cl_mem` attributes.
```

Comment:
Does the latter preference actually apply to OpenCL and (most of) its implementations, or only to the FPGA Runtime for OpenCL?

Reply:
This item is quite long and the order of calls cumbersome to read. How about removing the entire sentence "Specifically …" in favour of a link to the section below? Then expand the section below to include a second list of calls in the order preferred by the FPGA Runtime for OpenCL.